presto-main/src/main/java/com/facebook/presto/operator/aggregation/FixedDoubleHistogram.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.operator.aggregation;
+
+import io.airlift.slice.SizeOf;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.SliceOutput;
+import io.airlift.slice.Slices;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class FixedDoubleHistogram
+        implements Iterable<FixedDoubleHistogram.Bin>
+{
+    private static final byte FORMAT_TAG = 0;
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(FixedDoubleHistogram.class).instanceSize();
+
+    private final int bucketCount;
+    private final double min;
+    private final double max;
+    private final double[] weights;
+
+    public class Bin
+    {
+        public final double left;
+        public final double right;
+        public final double weight;
+
+        public Bin(double left, double right, double weight)
+        {
+            this.left = left;
+            this.right = right;
+            this.weight = weight;
+        }
+    }
+
+    public FixedDoubleHistogram(int bucketCount, double min, double max)
+    {
+        checkArgument(bucketCount >= 2, "bucketCount %s must be >= 2", bucketCount);
+        checkArgument(min < max, "min %s must be greater than max %s ", min, max);
+
+        this.bucketCount = bucketCount;
+        this.min = min;
+        this.max = max;
+        this.weights = new double[bucketCount];
+    }
+
+    public FixedDoubleHistogram(SliceInput input)
+    {
+        checkArgument(input.readByte() == FORMAT_TAG, "Unsupported format tag");
+        bucketCount = input.readInt();
+        checkArgument(bucketCount >= 2, "bucketCount %s must be >= 2", bucketCount);
+        min = input.readDouble();
+        max = input.readDouble();
+        checkArgument(min < max, "min %s must be greater than max %s ", min, max);
+        weights = new double[bucketCount];
+        input.readBytes(Slices.wrappedDoubleArray(weights), bucketCount * SizeOf.SIZE_OF_DOUBLE);
+    }
+
+    int getBucketCount()
+    {
+        return bucketCount;
+    }
+
+    double getMin()
+    {
+        return min;
+    }
+
+    double getMax()
+    {
+        return max;
+    }
+
+    FixedDoubleHistogram(FixedDoubleHistogram other)
```

Comment:
I feel that we normally don't do this in Presto but explicitly create a new object with old parameters.

Reply:
@rongrong I've made the class implement `Cloneable`, and implemented `clone`. This is now used instead.